### PR TITLE
Feat: File version control and helper methods

### PR DESCRIPTION
**Pull request by nidishk**
🔖 Title
Implement full version control support in FileManager

📝 Description
This pull request adds end-to-end versioning capabilities to FileManagerBase, including:

1. New versioned uploads via uploadWithVersioning, which transparently records each file’s batch, hash, size, timestamp, and operation (CREATE/MODIFY/DELETE) into a Swarm feed.
2. Manual versioning APIs:

- createFileVersion / deleteFileVersion to record updates and deletions on demand.
- getFileVersionHistoryByPath / getFileVersionInfoByPath / getFileVersion to query history, summary info, or individual entries.
- downloadFileVersion to fetch the exact bytes of any recorded version.
- hasFileVersionHistory and getVersionedFiles for quick existence checks and listing.

3. Low-level utilities in versionControl.ts: feed topic generation, count calculation, metadata read/write, content-hashing helper, and history scanning.
4. Comprehensive unit tests in fileManagerVersionControl.spec.ts covering success, failure, and edge cases.
5. Integration tests in versionControl.integration.spec.ts demonstrating a full lifecycle on a real Bee node.
6. Refactored error handling to surface uninitialized or missing-stamp conditions gracefully.
7. Minor updates to existing upload/download workflows to wire in version metadata where appropriate.

🔗 Related Issues
https://solar-punk.atlassian.net/browse/SPDV-469

🧪 How Has This Been Tested?
✅ Manually tested with a mocked Bee client 

✅ Checklist
✅ My code follows the project’s style guide
✅ I have performed a self-review of my code
✅ I have commented my code where necessary
✅ I have added tests that prove my fix/feature works
✅ All new and existing tests pass locally

### CODE REVIEW

**Review by bosi95 — CHANGES_REQUESTED** [review]
please rework the versioning with this.ownerFeedList and saving the feed index for each upload. See my comment above

File: src/fileManager.ts [hunk 1]
```diff
@@ -311,6 +313,9 @@ export class FileManagerBase implements FileManager {
     uploadOptions?: RedundantUploadOptions | FileUploadOptions | CollectionUploadOptions,
     requestOptions?: BeeRequestOptions,
   ): Promise<void> {
+    console.debug(`[FM] upload called for ${infoOptions.name} — infoOptions.customMetadata =`, infoOptions.customMetadata);
```
Comment (bosi95) [inline src/fileManager.ts:316]: unnecessary duplicate logs, please remove at least one of them

File: src/fileManager.ts [hunk 2]
```diff
@@ -311,6 +313,9 @@ export class FileManagerBase implements FileManager {
...
+    const tag = infoOptions.customMetadata;
+    console.debug(`[FM] upload called for ${infoOptions.name} — infoOptions.customMetadata =`, tag);
```
Comment (bosi95) [inline src/fileManager.ts:318]: Please use the same logging conventions, [FM] prefix is a good idea, but in that case please modify all the other logs

File: src/fileManager.ts [hunk 3]
```diff
@@ -311,6 +313,9 @@ export class FileManagerBase implements FileManager {
     uploadOptions?: RedundantUploadOptions | FileUploadOptions | CollectionUploadOptions,
     requestOptions?: BeeRequestOptions,
   ): Promise<void> {
+    console.debug(`[FM] upload called for ${infoOptions.name} — infoOptions.customMetadata =`, infoOptions.customMetadata);
+    const tag = infoOptions.customMetadata;
```
Comment (bosi95) [inline src/fileManager.ts:317]: I don't think a separate variable is needed for metadata.

File: src/fileManager.ts [hunk 4]
```diff
@@ -348,14 +360,41 @@ export class FileManagerBase implements FileManager {
       shared: false,
       preview: undefined,
       index: infoOptions.index,
-      customMetadata: infoOptions.customMetadata,
+      customMetadata: tag,
       redundancyLevel: uploadOptions?.redundancyLevel,
     };
+    console.debug(`[FM] constructed fileInfo.customMetadata =`, fileInfo.customMetadata);
```
Comment (bosi95) [inline src/fileManager.ts:366]: remove log

File: src/fileManager.ts [hunk 5]
```diff
@@ -348,14 +360,41 @@ export class FileManagerBase implements FileManager {
       shared: false,
       preview: undefined,
       index: infoOptions.index,
```
Comment (bosi95) [inline src/fileManager.ts:362]: use the index property here as versioning, we don't need a separate structure for it

File: src/fileManager.ts [hunk 6]
```diff
@@ -348,14 +360,41 @@ export class FileManagerBase implements FileManager {
...
     await this.saveFileInfoAndFeed(fileInfo, requestOptions);
 
-    this.emitter.emit(FileManagerEvents.FILE_UPLOADED, { fileInfo });
+    let version: number | undefined;
+    try {
+      version = await this.recordVersion(fileInfo);
+    } catch (err: any) {
+      console.warn(`⚠️  Failed to record version for ${fileInfo.name}:`, err);
+    }
+    console.log('[FileManager] recordVersion returned →', version);
```
Comment (bosi95) [inline src/fileManager.ts:375]: please use debug logs and the same convention

File: src/fileManager.ts [hunk 7]
```diff
@@ -348,14 +360,41 @@ export class FileManagerBase implements FileManager {
...
       index: infoOptions.index,
-      customMetadata: infoOptions.customMetadata,
+      customMetadata: tag,
       redundancyLevel: uploadOptions?.redundancyLevel,
     };
+    console.debug(`[FM] constructed fileInfo.customMetadata =`, fileInfo.customMetadata);
     await this.saveFileInfoAndFeed(fileInfo, requestOptions);
 
-    this.emitter.emit(FileManagerEvents.FILE_UPLOADED, { fileInfo });
+    let version: number | undefined;
```
Comment (bosi95) [inline src/fileManager.ts:369]: you don't need to separately record the versions as it is already stored in the (feed)index property.
If infoTopic is not provided then index can be regarded as the first (0th). 
If infoTopic and index is provided then just increase the index. 
If infoTopic is provided but index is not, then you can look up the feed of infoTopic and determine the index

File: src/utils/types.ts [hunk 8]
```diff
@@ -117,6 +132,23 @@ export interface FileManager {
   emitter: EventEmitter;
 }
 
+export interface FileVersionMetadata {
```
Comment (bosi95) [inline src/utils/types.ts:135]: not needed IF, already present in FileInfo

File: src/utils/types.ts [hunk 9]
```diff
@@ -130,6 +162,8 @@ export interface FileInfo {
   index?: string | undefined;
   redundancyLevel?: RedundancyLevel;
   customMetadata?: Record<string, string>;
+  versionInfo?: FileVersionInfo;
+  isVersioned?: boolean;
```
Comment (bosi95) [inline src/utils/types.ts:166]: remove   isVersioned

File: src/utils/types.ts [hunk 10]
```diff
@@ -130,6 +162,8 @@ export interface FileInfo {
   index?: string | undefined;
   redundancyLevel?: RedundancyLevel;
   customMetadata?: Record<string, string>;
+  versionInfo?: FileVersionInfo;
```
Comment (bosi95) [inline src/utils/types.ts:165]: I don't think you need to store it within the FileInfo

File: src/utils/types.ts [hunk 11]
```diff
@@ -117,6 +132,23 @@ export interface FileManager {
...
+  filePath: string;
+  contentHash: string;
+  size: number;
+  timestamp: string;
+  version: number;
+  batchId: string;
+  customMetadata?: Record<string, string>;
+}
+
+export interface FileVersionInfo {
```
Comment (bosi95) [inline src/utils/types.ts:145]: I think it is and IF unnecessary, a subset of FileInfo.

File: src/utils/types.ts [hunk 12]
```diff
@@ -192,3 +226,14 @@ export interface WrappedUploadResult {
   uploadFilesRes: Reference | string;
   uploadPreviewRes?: Reference | string;
 }
+
+export interface FileVersionMetadata {
```
Comment (bosi95) [inline src/utils/types.ts:230]: duplicate IF, please remove

File: src/utils/common.ts [hunk 13]
```diff
@@ -50,6 +51,15 @@ export function generateTopic(): Topic {
   return new Topic(getRandomBytesBrowser(Topic.LENGTH));
 }
 
+export function generateFileFeedTopic(filePath: string): Topic {
```
Comment (bosi95) [inline src/utils/common.ts:54]: no need for separate file version feed

File: src/fileManager.ts [hunk 14]
```diff
@@ -465,6 +504,67 @@ export class FileManagerBase implements FileManager {
     }
   }
 
+  async getVersionCount(filePath: string): Promise<number> {
```
Comment (bosi95) [inline src/fileManager.ts:507]: this function can work as I described it above.

File: src/fileManager.ts [hunk 15]
```diff
@@ -465,6 +504,67 @@ export class FileManagerBase implements FileManager {
...
+    const owner = this.signer.publicKey().address().toString();
+    const topic = generateFileFeedTopic(filePath);
+    const { feedIndexNext } = await getFeedData(this.bee, topic, owner);
+    // If feedIndexNext is undefined, treat it as zero
+    return Number(feedIndexNext?.toBigInt() ?? 0n);
+  }
+
+  async getVersion(
```
Comment (bosi95) [inline src/fileManager.ts:515]: this function can be simplified and changed:
instead of filepath use a fieinfo object as an argument coming from the client.
find the corresponding info in this.fileInfoList and return the feed and the specified index/ version.

File: src/fileManager.ts [hunk 16]
```diff
@@ -465,6 +504,67 @@ export class FileManagerBase implements FileManager {
...
+    filePath: string,
+    version: number,
+  ): Promise<FileVersionMetadata | null> {
+    const owner = this.signer.publicKey().address().toString();
+    const topic = generateFileFeedTopic(filePath);
+    let feedData;
+    try {
+      feedData = await getFeedData(this.bee, topic, owner, BigInt(version));
+    } catch (err: any) {
+      if (err.status === 404) {
```
Comment (bosi95) [inline src/fileManager.ts:525]: getFeedData already handles 404

File: src/fileManager.ts [hunk 17]
```diff
@@ -465,6 +504,67 @@ export class FileManagerBase implements FileManager {
...
+    }
+  
+    // if zero-address payload, nothing was ever written
+    if (SWARM_ZERO_ADDRESS.equals(feedData.payload.toUint8Array())) {
+      return null;
+    }
+  
+    // payload is the bytes of a JSON blob { reference, historyRef? }
+    const raw = feedData.payload.toUint8Array();
+    const text = Buffer.from(raw).toString('utf-8');
```
Comment (bosi95) [inline src/fileManager.ts:538]: you can use fileinfo assertion just like in initFileInfoList line 255

File: src/utils/types.ts [hunk 18]
```diff
@@ -99,6 +99,21 @@ export interface FileManager {
    */
   getGrantees(fileInfo: FileInfo): Promise<GetGranteesResult>;
 
+  /**
+   * How many versions exist for this file (0 = none).
+   */
+  getVersionCount(filePath: string): Promise<number>;
```
Comment (bosi95) [inline src/utils/types.ts:105]: rename filepath to fileinfo and use the topic in it
consider removing it, does the client need it ?
also 0 shall mean the first one, and -1 is none

File: src/utils/types.ts [hunk 19]
```diff
@@ -99,6 +99,21 @@ export interface FileManager {
...
+
+  /**
+   * Load a single version's metadata for a file.
+   */
+  getVersion(filePath: string, version: number): Promise<FileVersionMetadata | null>;
+
+  /**
+   * Load every version's metadata, in order.
+   */
+  getHistory(filePath: string): Promise<FileVersionMetadata[]>;
```
Comment (bosi95) [inline src/utils/types.ts:115]: remove getHistory, client can use getVersion multiple times with its desired versions.

File: src/utils/types.ts [hunk 20]
```diff
@@ -99,6 +99,21 @@ export interface FileManager {
...
 
+  /**
+   * How many versions exist for this file (0 = none).
+   */
+  getVersionCount(filePath: string): Promise<number>;
+
+  /**
+   * Load a single version's metadata for a file.
+   */
+  getVersion(filePath: string, version: number): Promise<FileVersionMetadata | null>;
```
Comment (bosi95) [inline src/utils/types.ts:110]: rename filepath to fileinfo and use the topic in it
return fileinfo | null

**Review by bosi95 — CHANGES_REQUESTED** [review]

**Comment by bosi95** [comment]
deprecated, please close this PR
